utility_scripts/tests/test_balance_dataset.py

```diff
@@ -0,0 +1,488 @@
+        balancer = Balancer(VerbosityPrinter())
+        balancer._set_bins(test_set['gnomAD_AF'])
+        consequences = balancer._obtain_consequences(test_set['Consequence'])
+        for consequence in consequences:
+            observed = balancer._process_consequence(
+                test_set[
+                    (test_set['Consequence'].str.contains(consequence, regex=False)) &
+                    (test_set['binarized_label'] == 1)
+                    ],
+                test_set[
+                    (test_set['Consequence'].str.contains(consequence, regex=False)) &
+                    (test_set['binarized_label'] == 0)
+                    ]
+            )
+            self.assertEqual(
+                observed.shape[0],
+                expected_rows[consequence],
+                msg=f'Test failed on consequence: {consequence}'
+            )
+
+    def test_balancer_known_input(self):
+        """
+        Tests the balancer function with a hardcoded dataframe
+        """
+        variant_data = [
+            ['variant_1', 'consequence_1', 0.8, 0],  # Paired with variant_2
+            ['variant_2', 'consequence_1', 0.75, 1],  # Paired with variant_1
+            ['variant_3', 'consequence_2', 0.2, 1],  # Paired with variant_5
+            ['variant_4', 'consequence_2', 0.01, 0],  # Paired with variant_6
+            ['variant_5', 'consequence_2', 0.45, 0],  # Paired with variant_3
+            ['variant_6', 'consequence_1&consequence_2', 0.02, 1],  # Paired with variant_4
+            ['variant_7', 'consequence_2&consequence_3', 0.001, 1],  # Removed
+            ['variant_8', 'consequence_3', 0.9, 0]  # Removed
+        ]
+        test_case = pd.DataFrame(
+            variant_data, columns=['variant', 'Consequence', 'gnomAD_AF', 'binarized_label']
+        )
+        balanced, _ = Balancer(VerbosityPrinter()).balance(test_case)
+        expected_variants = [
+            'variant_1',
+            'variant_2',
+            'variant_3',
+            'variant_4',
+            'variant_5',
+            'variant_6'
+        ]
+        for variant in balanced['variant'].values:
+            self.assertIn(variant, expected_variants)
+        self.assertFalse(balanced['variant'].duplicated().any())
+
+    def test_duplicate_processing_balancer(self):
+        """
+        Test to see if a sample with multiple consequences gets sampled only once
+        """
+        test_dataset = pd.DataFrame(
+            {
+                'Consequence': [
+                    'consequence_1&consequence_2',
+                    'consequence_1',
+                    'consequence_2',
+                    'consequence_2'
+                ],
+                'binarized_label': [0, 1, 0, 1],
+                'gnomAD_AF': [0.0, 0.0, 0.0, 0.0]
+            }
+        )
+        observed, _ = Balancer(VerbosityPrinter()).balance(test_dataset)
+        self.assertFalse(observed.duplicated().any())
```

Comment:
The second test will not include the call to `_test_consequence`. The limitation of `_process_consequence()` is that it does not check itself if a sample has been sampled already, that might make the method too complex or too computational expensive. The `balance()` method does check if a sample has been sampled already, since that method contains the loop through all consequences.

Reply:
Noted. Main point was to ensure test are present that show expected behavior when an item can be used for multiple consequences, how it would influence the number of items per consequence. Would be interesting to see a more complex scenario as well, though maybe something to think about for in the refactor. I see there's a test that validates that consequence 2 is not present in observed so this can be resolved.

Reply:
Noted. The test `test_process_consequence_multiconsequence` should cover for a more complex usecase, although adding more complex test cases might negatively influence the test readability.